core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1162,9 +1162,14 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
     @throws[Exception]
     def handleNewSession() {
       info("ZK expired; shut down all controller components and try to re-elect")
-      onControllerResignation()
-      inLock(controllerContext.controllerLock) {
-        controllerElector.elect
+      if (controllerElector.getControllerID() != config.brokerId) {
+        onControllerResignation()
+        inLock(controllerContext.controllerLock) {
+          controllerElector.elect
+        }
+      } else {
+        //maybe create by current session or the previous zk session's ephemeral node is not deleted
```

Comment:
@pengwei-li : I pushed a minor change to the comment. Please see if it looks good to you. For case 2, ZK server doesn't guarantee ephemeral nodes are deleted before expiring a session? I am getting conflicting info on this. So not sure.